File: server/src/routes/messages.ts
import { Router, Request, Response } from 'express'
import { authService } from '../services/authService'
import { 
  sendMessage, 
  getConversation, 
  getConversations, 
  getUnreadCount, 
  markAsRead, 
  markAllAsRead,
  deleteMessage 
} from '../services/messageService'

const router = Router()

// 发送消息
router.post('/send', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const { receiverId, receiverUsername, content } = req.body
    
    if (!receiverId || !receiverUsername || !content) {
      return res.status(400).json({ error: '缺少必要字段' })
    }
    
    if (!content.trim()) {
      return res.status(400).json({ error: '消息内容不能为空' })
    }
    
    // 不能给自己发消息
    if (receiverId === currentUser.id) {
      return res.status(400).json({ error: '不能给自己发消息' })
    }
    
    const message = await sendMessage(
      currentUser.id,
      currentUser.username,
      currentUser.avatar,
      receiverId,
      receiverUsername,
      content
    )
    
    res.status(201).json(message)
  } catch (error: any) {
    res.status(400).json({ error: error.message || '发送消息失败' })
  }
})

// 获取与某个用户的对话
router.get('/conversation/:userId', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const { userId } = req.params
    const limit = parseInt(req.query.limit as string) || 50
    const skip = parseInt(req.query.skip as string) || 0
    
    const messages = await getConversation(currentUser.id, userId, limit, skip)
    
    res.json(messages)
  } catch (error: any) {
    res.status(400).json({ error: error.message || '获取对话失败' })
  }
})

// 获取所有对话列表
router.get('/conversations', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const conversations = await getConversations(currentUser.id)
    
    res.json(conversations)
  } catch (error: any) {
    res.status(400).json({ error: error.message || '获取对话列表失败' })
  }
})

// 获取未读消息数
router.get('/unread', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const count = await getUnreadCount(currentUser.id)
    
    res.json({ count })
  } catch (error: any) {
    res.status(400).json({ error: error.message || '获取未读消息数失败' })
  }
})

// 标记与某个用户的消息为已读
router.put('/read/:userId', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const { userId } = req.params
    
    const count = await markAsRead(currentUser.id, userId)
    
    res.json({ count })
  } catch (error: any) {
    res.status(400).json({ error: error.message || '标记已读失败' })
  }
})

// 标记所有消息为已读
router.put('/read-all', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const count = await markAllAsRead(currentUser.id)
    
    res.json({ count })
  } catch (error: any) {
    res.status(400).json({ error: error.message || '标记已读失败' })
  }
})

// 删除消息
router.delete('/:messageId', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '')
    
    if (!token) {
      return res.status(401).json({ error: '未登录' })
    }
    
    const currentUser = await authService.verifyToken(token)
    if (!currentUser) {
      return res.status(401).json({ error: 'Token无效' })
    }
    
    const { messageId } = req.params
    
    const success = await deleteMessage(messageId, currentUser.id)
    
    if (!success) {
      return res.status(404).json({ error: '消息不存在' })
    }
    
    res.json({ success: true })
  } catch (error: any) {
    res.status(400).json({ error: error.message || '删除消息失败' })
  }
})

export default router

File: server/handlers/messages.go
package handlers

import (
	"net/http"
	"strconv"

	"blog-server/models"
	"blog-server/services"

	"github.com/gin-gonic/gin"
)

// SendMessage 发送消息
func SendMessage(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	var input models.SendMessageInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少必要字段"})
		return
	}

	msg, err := services.SendMessage(
		currentUser.ID,
		currentUser.Username,
		currentUser.Avatar,
		input.ReceiverID,
		input.ReceiverUsername,
		input.Content,
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "发送消息失败"})
		return
	}

	c.JSON(http.StatusCreated, msg)
}

// GetConversation 获取对话
func GetConversation(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	otherUserID := c.Param("userId")
	limitStr := c.DefaultQuery("limit", "50")
	skipStr := c.DefaultQuery("skip", "0")

	limit, _ := strconv.Atoi(limitStr)
	skip, _ := strconv.Atoi(skipStr)

	messages, err := services.GetConversation(currentUser.ID, otherUserID, limit, skip)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取对话失败"})
		return
	}

	c.JSON(http.StatusOK, messages)
}

// GetConversations 获取所有对话列表
func GetConversations(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	conversations, err := services.GetConversations(currentUser.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取对话列表失败"})
		return
	}

	c.JSON(http.StatusOK, conversations)
}

// GetUnreadCount 获取未读消息数
func GetUnreadCount(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	count, err := services.GetUnreadCount(currentUser.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取未读消息数失败"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"unreadCount": count})
}

// MarkAsRead 标记消息为已读
func MarkAsRead(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	senderID := c.Param("userId")

	_, err := services.MarkMessageAsRead(currentUser.ID, senderID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "标记已读失败"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

// MarkAllAsRead 标记所有消息为已读
func MarkAllAsRead(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	_, err := services.MarkAllMessagesAsRead(currentUser.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "标记已读失败"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

// DeleteMessage 删除消息
func DeleteMessage(c *gin.Context) {
	currentUser := getCurrentUser(c)
	if currentUser == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未登录"})
		return
	}

	messageID := c.Param("messageId")

	success, err := services.DeleteMessage(messageID, currentUser.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "删除消息失败"})
		return
	}

	if !success {
		c.JSON(http.StatusNotFound, gin.H{"error": "消息不存在"})
		return
	}

	c.Status(http.StatusNoContent)
}
